rrn.1
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.42)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "RRN 1"
.TH RRN 1 "2012-06-19" "rrn v1.0.0" "Regular Expression Rename"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
rrn \- Regular expression ReName
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBrrn\fR [\fIoptions\fR] \fI\s-1FROM\s0\fR [\fI\s-1TO\s0\fR] [\fI\s-1FILES\s0\fR | \fI\f(CI@LISTFILE\fI\fR]
.PP
\&\fBrrn\fR [\fIoptions\fR] [\fB\-c\fR|\fB\-f\fR] \fI\s-1CMD\s0\fR [\fI\s-1FILES\s0\fR | \fI\f(CI@LISTFILE\fI\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBrrn\fR is a program which will rename a set of files using
Perl-style regular expressions to specify how files should be
renamed.
.PP
The general syntax is as follows:
.Sp
.RS 4
\&\fBrrn\fR \fI\s-1FROM\s0\fR \fI\s-1TO\s0\fR [\fI\s-1FILES\s0\fR | \fI\f(CI@LISTFILE\fI\fR]
.RE
.PP
where \fI\s-1FROM\s0\fR is the part of a filename that should be changed, \fI\s-1TO\s0\fR is
what it should be changed to, and [\fI\s-1FILES\s0\fR] is an optional list of files
to process.  If no files are specified, all files in the current
directory are processed.
.PP
Optionally, instead of (or in addition to) a list of \fI\s-1FILES\s0\fR, you can also
specify a \fI\f(CI@LISTFILE\fI\fR, where \fI\s-1LISTFILE\s0\fR is a text file containing the list
of files to rename, one per line. Use the \fB\-@\fR option to inhibit this 
behavior.
.PP
The simplest instance is renaming all instances of a word in a
filename to a different word.  For instance:
.Sp
.RS 4
\&\f(CW\*(C`rrn .htm .html\*(C'\fR
.RE
.PP
is a quick and dirty way to rename files with the extension \fI.htm\fR to
files with an extension of \fI.html\fR.  In this case it is very similar
to the \fBrename\fR\|(1) command found in the util-linux package.
.PP
You can also use perl-style regular expressions to specify the text
that should be changed:
.Sp
.RS 4
\&\f(CW\*(C`rrn \*(Aq\es+\*(Aq \*(Aq \*(Aq\*(C'\fR
.RE
.PP
will replace strings of multiple spaces in a filename with a single
space.  Another example might look like the following:
.Sp
.RS 4
\&\f(CW\*(C`rrn \*(Aq([a\-z])([A\-Z])\*(Aq \*(Aq$1 $2\*(Aq\*(C'\fR
.RE
.PP
which would insert a space in between any lowercase letter followed by
an upper case letter, thus changing a name like
\&\fIDigitalCameraPictures\fR to \fIDigital Camera Pictures\fR.
.PP
If you're unfamiliar with Perl-style regexps, have a look at the
\&\fBperlre\fR\|(1), \fBperlrequick\fR\|(1), and/or \fBperlretut\fR\|(1) manpages that come with
Perl.
.PP
For those that are familiar with Perl, this script basically runs each
file name through an \f(CW\*(C`s/\f(CIFROM\f(CW/\f(CITO\f(CW/g\*(C'\fR command, and renaming it to
whatever comes out.  Obviously, if the substitution results in the
same name as the original, the file will not be renamed.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-t\fR" 4
.IX Item "-t"
Test.  Only shows what changes would be made, without actually making
them
.IP "\fB\-a\fR" 4
.IX Item "-a"
Ask.  Shows the changes that are about to be made, then ask whether to
apply them
.IP "\fB\-v\fR" 4
.IX Item "-v"
Verbose.  Shows the names of files and what they're renamed to as it's
renaming them.  By default \fBrrn\fR produces no output unless an error
occurs
.IP "\fB\-g\fR" 4
.IX Item "-g"
No global.  Applies the substitution to only the first occurance of
\&\fI\s-1FROM\s0\fR.  This is like removing the /g from the end of the s// regexp
in Perl.
.IP "\fB\-i\fR" 4
.IX Item "-i"
Case insensitive.  Letters will be matched against the \fI\s-1FROM\s0\fR pattern
without regards to capitality.  Note that the matched portion of the
name will still be replaced with the \fI\s-1TO\s0\fR text exactly as you
specified it.  Therefore, \f(CW\*(C`rrn \-i a b\*(C'\fR will change a
capital 'A' to a lowercase 'b'. This is like adding an /i to the end
of the s// operator
.IP "\fB\-e\fR" 4
.IX Item "-e"
Don't change the file's extension.  (Extension is defined as the
portion of the filename including and following the final '.' character in
the file, if there is one)  This removes the extension before substitution,
and replaces it after substitution, so your \fI\s-1FROM\s0\fR pattern won't even
match it.
.IP "\fB\-c\fR \fI\s-1CMD\s0\fR" 4
.IX Item "-c CMD"
Perform \fI\s-1CMD\s0\fR, which is a Perl expression.  If this option is
specified, the \fI\s-1FROM\s0\fR and \fI\s-1TO\s0\fR parameters must be left off.  The
original file name will be found in the \f(CW$_\fR variable; your command
should in some fashion modify that variable to what you want the new
name to be.  \f(CW\*(C`rrn \-c \*(Aqs/FROM/TO/g\*(Aq\*(C'\fR is equivalent to
\&\f(CW\*(C`rrn FROM TO\*(C'\fR
.Sp
Note that if the \fB\-e\fR option is used, the extension will be removed
from the \f(CW$_\fR variable.  This way you can append something to the end of
the name, and it will appear between the end of the filename and the
start of the extension.
.IP "\fB\-f\fR \fI\s-1CMD\s0\fR" 4
.IX Item "-f CMD"
Similar to \fB\-c\fR, but the result of the evaluated expression is used as the new
name.  \f(CW\*(C`rrn \-f lc\*(C'\fR is a quick way to convert a filename to all-lowercase
characters.  (Or use \f(CW\*(C`uc\*(C'\fR for uppercase.) As with \fB\-c\fR, the original filename
is found in \f(CW$_\fR, and whatever function you call should access that variable.
Many Perl builtin functions will implicitely acecss \f(CW$_\fR, so you can simply say,
for instance, \f(CW\*(C`lc\*(C'\fR instead of \f(CW\*(C`lc($_)\*(C'\fR.
.Sp
Note that this doesn't have to be an actual function, it can be any
expression that returns a value.  So you can say:
.RS 4
.Sp
.RS 4
\&\f(CW\*(C`rrn \-f \*(Aq$foo++ . "_$_"\*(Aq\*(C'\fR
.RE
.RE
.RS 4
.Sp
to prepend a number before the filename.
.RE
.IP "\fB\-E\fR" 4
.IX Item "-E"
A distant cousin to \fB\-f\fR and \fB\-c\fR, if this is specified then the \fI\s-1FROM\s0\fR
parameter is a regex (as usual), but the \fI\s-1TO\s0\fR parameter is a Perl expression,
not a regex. Thus, it is somewhat of a hybrid between the \fB\-f\fR/\fB\-c\fR options
and the defualt behaviour.  Thus,
.RS 4
.Sp
.RS 4
\&\f(CW\*(C`rrn \-E \*(Aq(^\ed)\*(Aq \*(Aq$1 + 1\*(Aq\*(C'\fR
.RE
.RE
.RS 4
.Sp
will increment the number at the start of the filename.
.Sp
This option is equivalent to \f(CW\*(C`rrn FROM \*(Aq${\e(TO)}\*(Aq\*(C'\fR.
.RE
.IP "\fB\-b\fR \fI\s-1INIT\s0\fR" 4
.IX Item "-b INIT"
Evaluates the expression \fI\s-1INIT\s0\fR before processing any files.  This is
useful if you want to reference a variable in your command, but need
it to be initialized to something first.  For instance:
.RS 4
.Sp
.RS 4
\&\f(CW\*(C`rrn \-b \*(Aq$foo = 1\*(Aq \-f \*(Aq$foo++ . "_$_"\*(Aq\*(C'\fR
.RE
.RE
.RS 4
.Sp
will prepend a number and an underscore to the filename.
.RE
.IP "\fB\-F\fR" 4
.IX Item "-F"
Force: Renames the file even if a file with the new name already exists.
\&\fBUse with caution!\fR
.IP "\fB\-@\fR" 4
.IX Item "-@"
This will prevent \fBrrn\fR from treating files with names starting with an @
symbol as list files.
.SH "DIAGNOSTICS"
.IX Header "DIAGNOSTICS"
.IP "Skipped \fI\s-1FILE\s0\fR: \fI\s-1NAME\s0\fR already exists" 4
.IX Item "Skipped FILE: NAME already exists"
A file wasn't renamed because the filename it was to be
renamed to was already in use.
.IP "Could not rename \fI\s-1FILE\s0\fR: \fI\s-1ERROR\s0\fR" 4
.IX Item "Could not rename FILE: ERROR"
A file wasn't renamed because an \s-1OS\s0 error occured
.SH "BUGS"
.IX Header "BUGS"
Renaming a file to a name with different capitalization but otherwise the
same under Windows will complain about the file already existing.
.PP
Probably more, but I use this program all the time so the major ones should be
long since squashed.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBrename\fR\|(1), \fBperlre\fR\|(1), \fBperlrequick\fR\|(1), \fBperlretut\fR\|(1), \fBperlfunc\fR\|(1), and
other documentation referenced in \fBperl\fR\|(1)
.SH "AUTHOR"
.IX Header "AUTHOR"
Nathan Roberts <nroberts@tardislabs.com>
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright 2012 by Nathan Roberts <nroberts@tardislabs.com>
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the \s-1GNU\s0 General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.


rrn.pl
#!/usr/bin/perl

# rrn: Regular expression ReName
# Copyright (C) 2012  Nathan Roberts
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

use strict;
use Getopt::Std;

our $VERSION = "1.0.0";
my $DATE = "2012-06-19";

# Get command-line options and transfer them to meaningful variable names

my %opts;

getopts( '@tavgc:f:Eeib:Fh', \%opts );

my( $nolist, $test, $ask, $verbose, $noglobal, $cmd, $func, $expr, $casein,
    $keepext, $begin, $force, $help ) = @opts{ qw( @ t a v g c f E i e b F h )};

# Print help message if requested

if( $help )
{
    HELP_MESSAGE();
    exit;
}

# Grab the 'from' and 'to' portions (regex parameters) if appropriate

my( $from, $to );

if( !$cmd && !$func )
{
    $from = shift;
    $to = shift;
}

# die "Missing parameter\n" if( $from eq '' && $cmd eq '' && $func eq '' );
HELP_MESSAGE() if( $from eq '' && $cmd eq '' && $func eq '' );

# Some debugging output that really isn't necessary anymore
#
# if( $cmd )
# {
#     print "Command: $cmd\n";
# }
# elsif( $func )
# {
#     print "Function: $func\n";
# }
# else
# {
#     print "From: $from\nTo: $to\n";
# }

# Get the list of files to rename

my @dir = @ARGV;

if( @dir && $^O eq 'MSWin32' )
{
    # We have to glob our own wildcards under Win32
    @dir = <@dir>;
}

# Expand @lists

if( !$nolist )
{
    my @temp = @dir;
    @dir = ();

    foreach my $file (@temp)
    {
	if( $file =~ /@(.+)/ )
	{
	    open my $handle, $1 
		or die "Erorr reading $1: $@\n";
	    my @foo = <$handle>;
	    chomp @foo;
	    push @dir, @foo;
	    close $handle;
	}
	else
	{
	    push @dir, $file;
	}
    }
}

if( !@dir )
{
    opendir( DIR, "./" );
    @dir = sort readdir( DIR );
}

# Figure out regex flags

my $opts;

if( !$noglobal )
{
    $opts .= 'g';
}

if( $casein )
{
    $opts .= 'i';
}

# Execute the init statement, if given

if( $begin )
{
    no strict;
    eval $begin;
    use strict;
    die $@ if( $@ );
}

# Figure out new names

my @renames;

foreach (@dir)
{
    next if( /^\.{1,2}$/ ); # Skip . and .. directories

    my $old = $_;

    # Save the file extension (if requested) and the pathname

    my $ext;
    if( $keepext )
    {
	($_, $ext) = /(.*)(\..*)/ or $_ = $old;
    }
    # FIXME: We ought to be using dirname and basename for this
    my $path;
    ($path, $_) = m!(.*/)?([^/]+)!;

    # Determine the new name, depending on which mode we're in
    # Note that 'strict' is temporarily turned off so that the user
    # can use variables without having to declare them first

    if( $cmd )
    {
	no strict;
	eval $cmd;
	use strict;
	die $@ if( $@ );
    }
    elsif( $func )
    {
	no strict;
	$_ = eval $func;
	use strict;
	die $@ if( $@ );
    }
    else
    {
	# Escape forward slashes, just in case, even though they'll
	# probably never be used, since they're not valid in a filename

	$from =~ s!/!\\/!g;
	$to =~ s!/!\\/!g;

	no strict;
	if( $expr )
	{
	    eval "s/$from/\${\\($to)}/$opts";
	}
	else
	{
	    eval "s/$from/$to/$opts";
	}
	use strict;
	die $@ if( $@ );
    }

    # Restore the extension and path

    $_ .= $ext if( $keepext );
    $_ = "$path$_";

    # No-op if name hasn't changed

    next if( $old eq $_ || $_ eq '' );

    # Store the new name

    push @renames, [$old, $_];
}

# Do test, if necessary
if( $test || $ask )
{
    for my $rename (@renames)
    {
	my( $old, $new ) = @$rename;

	print "$old --> $new\n";
    }
}

exit if( $test );

if( $ask )
{
    print "Apply? [y/N] ";
    my $response = <STDIN>;
    if( $response !~ /^y/i )
    {
	print "Aborted.\n";
	exit;
    }
}

# Do the acual rename

for my $rename (@renames)
{
    my( $old, $new ) = @$rename;

    if( -e $new && !$force )
    {
	warn "Skipped $old: $new already exists\n";
	next;
    }
    elsif( !rename( $old, $new ))
    {
	warn "Could not rename $old: $!\n";
	next;
    }
    print "$old --> $_\n" if( $verbose );
}

sub HELP_MESSAGE
{
    print "
rrn: Regular expression ReName - v$VERSION ($DATE)

Syntax:
  rrn [-tavgieFE] [-b INIT] FROM TO [FILES | \@LISTFILE]
  rrn [-taveF] [-b INIT] [-c | -f] CMD [FILES | \@LISTFILE]

  -t     Test: Only show what changes would have been made
  -a     Ask: Show changes (like -t) then confirm
  -v     Verbose: Show the files as they are being renamed
  -g     No global: Apply substitution to only the first occurance
  -i     Ignore case: Pattern matching will be case-insensitive
  -e     Don't change file extension
  -b     Execute INIT expression before any substitutions
  -F     Force: Rename even if new name already exists.

  -c     Perform a command on filenames instead of substitution
  -f     Perform a function on filenames instead of substitution
  -E     TO is a perl expression instead of a regex parameter

  -@     Don't process \@LISTFILE arguments

";
exit;
}

sub VERSION_MESSAGE
{
    print "rrn v$VERSION\n";
    exit;
}

__END__

=head1 NAME

rrn - Regular expression ReName

=head1 SYNOPSIS

B<rrn> [I<options>] I<FROM> [I<TO>] [I<FILES> | I<@LISTFILE>]

B<rrn> [I<options>] [B<-c>|B<-f>] I<CMD> [I<FILES> | I<@LISTFILE>]

=head1 DESCRIPTION

B<rrn> is a program which will rename a set of files using
Perl-style regular expressions to specify how files should be
renamed.

The general syntax is as follows:

=over

B<rrn> I<FROM> I<TO> [I<FILES> | I<@LISTFILE>]

=back

where I<FROM> is the part of a filename that should be changed, I<TO> is
what it should be changed to, and [I<FILES>] is an optional list of files
to process.  If no files are specified, all files in the current
directory are processed.

Optionally, instead of (or in addition to) a list of I<FILES>, you can also
specify a I<@LISTFILE>, where I<LISTFILE> is a text file containing the list
of files to rename, one per line. Use the B<-@> option to inhibit this 
behavior.

The simplest instance is renaming all instances of a word in a
filename to a different word.  For instance:

=over

C<rrn .htm .html>

=back

is a quick and dirty way to rename files with the extension I<.htm> to
files with an extension of I<.html>.  In this case it is very similar
to the rename(1) command found in the util-linux package.

You can also use perl-style regular expressions to specify the text
that should be changed:

=over

C<rrn '\s+' ' '>

=back

will replace strings of multiple spaces in a filename with a single
space.  Another example might look like the following:

=over

C<rrn '([a-z])([A-Z])' '$1 $2'>

=back

which would insert a space in between any lowercase letter followed by
an upper case letter, thus changing a name like
I<DigitalCameraPictures> to I<Digital Camera Pictures>.

If you're unfamiliar with Perl-style regexps, have a look at the
perlre(1), perlrequick(1), and/or perlretut(1) manpages that come with
Perl.

For those that are familiar with Perl, this script basically runs each
file name through an C<s/I<FROM>/I<TO>/g> command, and renaming it to
whatever comes out.  Obviously, if the substitution results in the
same name as the original, the file will not be renamed.

=head1 OPTIONS

=over

=item B<-t>

Test.  Only shows what changes would be made, without actually making
them

=item B<-a>

Ask.  Shows the changes that are about to be made, then ask whether to
apply them

=item B<-v>

Verbose.  Shows the names of files and what they're renamed to as it's
renaming them.  By default B<rrn> produces no output unless an error
occurs

=item B<-g>

No global.  Applies the substitution to only the first occurance of
I<FROM>.  This is like removing the /g from the end of the s// regexp
in Perl.

=item B<-i>

Case insensitive.  Letters will be matched against the I<FROM> pattern
without regards to capitality.  Note that the matched portion of the
name will still be replaced with the I<TO> text exactly as you
specified it.  Therefore, C<rrn -i a b> will change a
capital 'A' to a lowercase 'b'. This is like adding an /i to the end
of the s// operator

=item B<-e>

Don't change the file's extension.  (Extension is defined as the
portion of the filename including and following the final '.' character in
the file, if there is one)  This removes the extension before substitution,
and replaces it after substitution, so your I<FROM> pattern won't even
match it.

=item B<-c> I<CMD>

Perform I<CMD>, which is a Perl expression.  If this option is
specified, the I<FROM> and I<TO> parameters must be left off.  The
original file name will be found in the $_ variable; your command
should in some fashion modify that variable to what you want the new
name to be.  C<rrn -c 's/FROM/TO/g'> is equivalent to
C<rrn FROM TO>

Note that if the B<-e> option is used, the extension will be removed
from the $_ variable.  This way you can append something to the end of
the name, and it will appear between the end of the filename and the
start of the extension.

=item B<-f> I<CMD>

Similar to B<-c>, but the result of the evaluated expression is used as the new
name.  C<rrn -f lc> is a quick way to convert a filename to all-lowercase
characters.  (Or use C<uc> for uppercase.) As with B<-c>, the original filename
is found in $_, and whatever function you call should access that variable.
Many Perl builtin functions will implicitely acecss $_, so you can simply say,
for instance, C<lc> instead of C<lc($_)>.

Note that this doesn't have to be an actual function, it can be any
expression that returns a value.  So you can say:

=over

C<rrn -f '$foo++ . "_$_"'>

=back

to prepend a number before the filename.

=item B<-E>

A distant cousin to B<-f> and B<-c>, if this is specified then the I<FROM>
parameter is a regex (as usual), but the I<TO> parameter is a Perl expression,
not a regex. Thus, it is somewhat of a hybrid between the B<-f>/B<-c> options
and the defualt behaviour.  Thus,

=over

C<rrn -E '(^\d)' '$1 + 1'>

=back

will increment the number at the start of the filename.

This option is equivalent to C<rrn FROM '${\(TO)}'>.

=item B<-b> I<INIT>

Evaluates the expression I<INIT> before processing any files.  This is
useful if you want to reference a variable in your command, but need
it to be initialized to something first.  For instance:

=over

C<rrn -b '$foo = 1' -f '$foo++ . "_$_"'>

=back

will prepend a number and an underscore to the filename.

=item B<-F>

Force: Renames the file even if a file with the new name already exists.
B<Use with caution!>

=item B<-@>

This will prevent B<rrn> from treating files with names starting with an @
symbol as list files.

=back

=head1 DIAGNOSTICS

=over

=item Skipped I<FILE>: I<NAME> already exists

A file wasn't renamed because the filename it was to be
renamed to was already in use.

=item Could not rename I<FILE>: I<ERROR>

A file wasn't renamed because an OS error occured

=back

=head1 BUGS

Renaming a file to a name with different capitalization but otherwise the
same under Windows will complain about the file already existing.

Probably more, but I use this program all the time so the major ones should be
long since squashed.

=head1 SEE ALSO

rename(1), perlre(1), perlrequick(1), perlretut(1), perlfunc(1), and
other documentation referenced in perl(1)

=head1 AUTHOR

Nathan Roberts <nroberts@tardislabs.com>

=head1 COPYRIGHT AND LICENSE

Copyright 2012 by Nathan Roberts <nroberts@tardislabs.com>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.

=cut
